src/ring_theory/int/basic.lean

```diff
@@ -341,3 +341,54 @@ instance decidable_int : decidable_rel (λ a b : ℤ, (multiplicity a b).dom) :=
 λ a b, decidable_of_iff _ finite_int_iff.symm
 
 end multiplicity
+
+lemma induction_on_primes {P : nat → Prop} (h₀ : P 0) (h₁ : P 1)
```

Comment:
I think this is non-trivial to actually do, but I'd be inclined to leave a breadcrumb for future contributors:
/-- TODO: generalize this to `ℕ → Sort*`, by iterating over `nat.factors`. -/
lemma induction_on_primes {P : ℕ → Prop} (h₀ : P 0) (h₁ : P 1)
```

Reply:
@semorrison, what are you thoughts on:
```lean
lemma induction_on_primes {P : nat → Sort*} (h₀ : P 0) (h₁ : P 1)
  (h : ∀ {p a : ℕ}, p.prime → P a → P (p * a)) : Π (n : ℕ), P n
| 0 := h₀
| 1 := h₁
| n@(k + 2) :=
  let m := n.min_fac in have n / m < n := nat.factors_lemma,
  have m.prime := nat.min_fac_prime $ nat.succ_injective.ne $ nat.succ_ne_zero _,
  have m * (n / m) = n := nat.mul_div_cancel' (nat.min_fac_dvd _),
  this.rec $ h ‹_› $ induction_on_primes _
```

Reply:
Seems nice.

In a slightly divergent direction, one could also replace `unique_factorization_domain.induction_on_irreducible` with
```
noncomputable 
def psigma_of_exists {α : Type*} {Z : α → Prop} (s : ∃ a, Z a) : Σ' a, Z a :=
⟨classical.some s, classical.some_spec s⟩

@[elab_as_eliminator] noncomputable def induction_on_irreducible {P : α → Sort*} (a : α)
  (h0 : P 0) (hu : ∀ u : α, is_unit u → P u)
  (hi : ∀ a i : α, a ≠ 0 → irreducible i → P a → P (i * a)) :
  P a :=
by haveI := classical.dec; exact
well_founded.fix wf_dvd_monoid.well_founded_dvd_not_unit
  (λ a ih, if ha0 : a = 0 then (congr_arg P ha0).mpr h0
    else if hau : is_unit a then hu a hau
    else let ⟨i, hii, B⟩ := psigma_of_exists (exists_irreducible_factor hau ha0) in
      let ⟨b, hb⟩ := psigma_of_exists B in
      have hb0 : b ≠ 0, from λ hb0, by simp * at *,
      (congr_arg P hb).mpr $ hi _ _ hb0 hii (ih _ ⟨hb0, i,
        hii.1, by rw [hb, mul_comm]⟩))
  a
```
(This can be cleaned up, I'm sure, but is irretrievably noncomputable.) I didn't try to write an equation for it.

Reply:
For your computable version on `nat`, what equations could we write for it? I fear that you're not going to be able to say `P (a * b) = P (b * a)`, but will have to have an explicit transport in there, involving `(congr_arg P (mul_comm a b)).mp ...`.

Reply:
(@thejohncrafter, this discussion is probably getting to be tangential to the actual PR; sorry for the noise).

Reply:
@eric-wieser @semorrison Do you still want to keep working on this or should I close this conversation ?

Reply:
> (@thejohncrafter, this discussion is probably getting to be tangential to the actual PR; sorry for the noise).

(No problem :) )

Reply:
I'm inclined to ignore this, and close this conversation. Neither @eric-wieser's or my suggestions results in a `Sort*` level definition that you can actually say anything about, so they are not really better than reducing to `Prop` via `nonempty`.